### DISCUSSION TITLE
Improvements for single tab and floating groups (constraints)

### DISCUSSION THREAD

**Discussion by Aventuum** [mathuo/dockview#405] [Ideas]
Dockview looks awesome and I'm seeing if it can replace another dock manager that I've already spent significant time hacking to make work (almost) the way I'd like.

For my use case I'm not using tabs, simply 3 panels initially stacked left to right: list, toolbar, content. Only the toolbar should be able to float, and it has a fixed width or height depending on if it's stacked horizontally or vertically. The toolbar orientation would be opposite to the parent splitview orientation, or user togglable if the toolbar panel is floating. In vertical mode, width is fixed to 150px and in horizontal mode the height is fixed to 150px.

Example: Initially, toolbar is in a horizontal splitview with list and content, so its orientation is vertical. It renders a vertical stack of toolbar buttons. If the toolbar is dragged above or below another panel, it gets rendered in a vertical splitview with that other panel, so the toolbar orientation changes to horizontal, and the toolbar buttons render in a horizontal stack. The user then floats the toolbar, it retains its last orientation, and a button can be used to toggle.

From my testing I think these features would be good:
- Get parent splitview orientation (maybe from the group api?)
- Drag to float without Shift key (float when dragging into an empty space, dock when dragging close to an edge as normal)
- Move floating window without Shift key (already achieved by dragging "void-container", but that isn't visible in singleTab mode)
- Disable floatable per panel
- Apply size constraints to floating panels

**Comment by mathuo** [comment]
Trying to understand this a little more...

So the "default" layout may be something like this:

![image](https://github.com/mathuo/dockview/assets/6710312/325b9958-2eda-4f6e-9f72-bbfba8132653)

where the "toolbar" panel can be freely dragged into new locations, if that location is "within another panel" then the orientation of the toolbar flips:

![image](https://github.com/mathuo/dockview/assets/6710312/cc453e1c-d716-4be0-b5a2-329c8081215e)

and then the other pieces are related to:
- if it's floating then the user can flip the orienation through some button (through an api method)
- the panels themselves may have fixed widths or heights in any mode (floating or fixed)

What do you mean by "disable floating per panel"?

**Reply by Aventuum** [reply]
Hi Mathuo, you've pretty much got it. By disable floating per panel, I mean each panel can have a property that determines whether that individual panel can be floated, if false then you can only drag to rearrange, not pop-out/float. In my example, the toolbar would be floatable, but the list and content would not be.

Here's my own mockups to slightly better show what I mean

![Default](https://github.com/mathuo/dockview/assets/22068586/09e74e4b-f298-469f-a8f6-a227985b404e)
![ToolbarFloatVertical](https://github.com/mathuo/dockview/assets/22068586/4ca74eda-ce0a-41bb-98b0-0ed427ce2608)
![ToolbarFloatHorizontal](https://github.com/mathuo/dockview/assets/22068586/8e85cd8a-59b2-46e8-9bee-95404c99ab28)
![AboveContent](https://github.com/mathuo/dockview/assets/22068586/45cc8112-4c22-460b-9ae8-6b21e676de74)
